### PR TITLE
Refactor log probability calculations into separate utility functions

### PR DESCRIPTION
**Pull request by hyeok9855**
See the issue #211.

To address the need for calculating log probabilities in `Sampler`, we decided to separate the logic for computing `log_probs` from the `GFlowNet` classes and instead create utility functions for these calculations.

**Naming (of variables, files, etc.) suggestions are welcomed!**

### CODE REVIEW

File: src/gfn/gflownet/detailed_balance.py [hunk 1]
```diff
@@ -78,6 +79,13 @@ def logF_parameters(self):
                 )
             )
 
+    def get_pfs_and_pbs(
```
Comment (josephdviviano) [inline src/gfn/gflownet/detailed_balance.py:82]: sorry, I misunderstood. I quite like this factorization.

File: src/gfn/utils/prob_calculations.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+    return log_pb_trajectories
+
+
+########################
+##### Transitions  #####
+########################
+
+
+def get_trans_pfs_and_pbs(
```
Comment (josephdviviano) [inline src/gfn/utils/prob_calculations.py:163]: sorry to be picky about this but I think we should avoid calling this `trans`. Maybe better to stick with `transitions`.

File: src/gfn/utils/prob_calculations.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+            trajectories.actions.tensor[..., 0],
+            fill_value=fill_value,
+            dtype=torch.float,
+        )
+        log_pf_trajectories[action_mask] = valid_log_pf_actions
+
+    return log_pf_trajectories
+
+
+def get_traj_pbs(
```
Comment (josephdviviano) [inline src/gfn/utils/prob_calculations.py:116]: For consistency, I'd also suggest we use here `trajectory` (ditto for all other instances)

File: src/gfn/utils/prob_calculations.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        with no_conditioning_exception_handler(module_name, module):
+            return module(states)
+
+
+#########################
+##### Trajectories  #####
+#########################
+
+
+def get_traj_pfs_and_pbs(
```
Comment (josephdviviano) [inline src/gfn/utils/prob_calculations.py:34]: `trajectory`

File: src/gfn/utils/prob_calculations.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    )
+    assert log_pb_trajectories.shape == (
+        trajectories.max_length,
+        trajectories.n_trajectories,
+    )
+
+    return log_pf_trajectories, log_pb_trajectories
+
+
+def get_traj_pfs(
```
Comment (josephdviviano) [inline src/gfn/utils/prob_calculations.py:68]: `trajectory`

File: src/gfn/utils/prob_calculations.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    log_pf_transitions = get_trans_pfs(pf, transitions, recalculate_all_logprobs)
+    log_pb_transitions = get_trans_pbs(pb, transitions)
+
+    assert log_pf_transitions.shape == (transitions.n_transitions,)
+    assert log_pb_transitions.shape == (transitions.n_transitions,)
+
+    return log_pf_transitions, log_pb_transitions
+
+
+def get_trans_pfs(
```
Comment (josephdviviano) [inline src/gfn/utils/prob_calculations.py:181]: `transitions`

File: src/gfn/utils/prob_calculations.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+        )
+
+        log_pf_actions = pf.to_probability_distribution(
+            states, estimator_outputs
+        ).log_prob(actions.tensor)
+
+    return log_pf_actions
+
+
+def get_trans_pbs(pb: GFNModule, transitions: Transitions) -> torch.Tensor:
```
Comment (josephdviviano) [inline src/gfn/utils/prob_calculations.py:206]: `transitions`

File: src/gfn/gflownet/detailed_balance.py [hunk 8]
```diff
@@ -78,6 +79,13 @@ def logF_parameters(self):
...
+        self, transitions: Transitions, recalculate_all_logprobs: bool = False
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        return get_trans_pfs_and_pbs(
```
Comment (josephdviviano) [inline src/gfn/gflownet/detailed_balance.py:85]: `transitions`

File: src/gfn/gflownet/detailed_balance.py [hunk 9]
```diff
@@ -101,70 +109,39 @@ def get_scores(
...
-                with has_conditioning_exception_handler("pf", self.pf):
-                    module_output = self.pf(states, transitions.conditioning)
-            else:
-                with no_conditioning_exception_handler("pf", self.pf):
-                    module_output = self.pf(states)
-
-            valid_log_pf_actions = self.pf.to_probability_distribution(
-                states, module_output
-            ).log_prob(actions.tensor)
+        log_pf_actions, log_pb_actions = self.get_pfs_and_pbs(
```
Comment (josephdviviano) [inline src/gfn/gflownet/detailed_balance.py:120]: ok I really like this.

**Review by josephdviviano — CHANGES_REQUESTED** [review]
This is an awesome PR, I think having these functions factored out will come in handy down the line as well for novel methods which might calculate these elements slightly differently. I think we should rename the function calls but this is a minor nit. I'll also wait for @younik  before approving.

File: src/gfn/utils/prob_calculations.py (hunk 2 above)
Comment (hyeok9855) [reply src/gfn/utils/prob_calculations.py:163]: Alright!

**Review by josephdviviano — APPROVED** [review]
lgtm!

File: src/gfn/gflownet/detailed_balance.py [hunk 10]
```diff
@@ -179,23 +156,21 @@ def get_scores(
             with no_conditioning_exception_handler("logF", self.logF):
                 valid_log_F_s_next = self.logF(valid_next_states).squeeze(-1)
 
-        targets[~valid_transitions_is_done] = valid_log_pb_actions
-        log_pb_actions = targets.clone()
-        targets[~valid_transitions_is_done] += valid_log_F_s_next
+        log_F_s_next = torch.zeros_like(log_pb_actions)
+        log_F_s_next[~valid_transitions_is_done] += valid_log_F_s_next
```
Comment (younik) [inline src/gfn/gflownet/detailed_balance.py:160]: with "=" is more clear

**Review by younik — APPROVED** [review]
Very good, it makes code more readable!

Thank you :)

**Pull request merged**